### PR TITLE
[cloud_firestore] Clean up Android Firestore listeners

### LINKED ISSUE #2390: Cloud Firestore Flutter java.lang.NullPointException
When I try to setData to a nested collection in cloud firestore , this error comes and forces my app to shut down but it happens only one particular collection and none other collection.

```
D/AndroidRuntime( 9744): Shutting down VM
E/AndroidRuntime( 9744): FATAL EXCEPTION: main
E/AndroidRuntime( 9744): Process: com.example.chatbot, PID: 9744
E/AndroidRuntime( 9744): java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
E/AndroidRuntime( 9744): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$EventObserver.onEvent(CloudFirestorePlugin.java:451)
E/AndroidRuntime( 9744): 	at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$EventObserver.onEvent(CloudFirestorePlugin.java:433)
E/AndroidRuntime( 9744): 	at com.google.firebase.firestore.Query.lambda$addSnapshotListenerInternal$2(com.google.firebase:firebase-firestore@@21.3.0:1025)
E/AndroidRuntime( 9744): 	at com.google.firebase.firestore.Query$$Lambda$3.onEvent(Unknown Source:6)
E/AndroidRuntime( 9744): 	at com.google.firebase.firestore.core.AsyncEventListener.lambda$onEvent$0(com.google.firebase:firebase-firestore@@21.3.0:42)
E/AndroidRuntime( 9744): 	at com.google.firebase.firestore.core.AsyncEventListener$$Lambda$1.run(Unknown Source:6)
E/AndroidRuntime( 9744): 	at android.os.Handler.handleCallback(Handler.java:883)
E/AndroidRuntime( 9744): 	at android.os.Handler.dispatchMessage(Handler.java:100)
E/AndroidRuntime( 9744): 	at android.os.Looper.loop(Looper.java:214)
E/AndroidRuntime( 9744): 	at android.app.ActivityThread.main(ActivityThread.java:7682)
E/AndroidRuntime( 9744): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 9744): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
E/AndroidRuntime( 9744): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
I/Process ( 9744): Sending signal. PID: 9744 SIG: 9
Lost connection to device.
```

### LINKED ISSUE #2517: [cloud_firestore] App Unexpectedlyclosed, shutting down VM, Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
**Describe the bug**
App closed unexpectedly while it was connected and idle. There was no user side interaction but there could be changes in the database which the app could be listening to.

**To Reproduce**
Could not reproduce this error, it seemed like a random event. Could not find out what triggered the crash.

**Expected behavior**
No such crash should have happened without triggering anything manually.

**Additional context**
Error logs
```
Shutting down VM
E/AndroidRuntime( 1822): FATAL EXCEPTION: main
E/AndroidRuntime( 1822): Process: com.piety.pietystore, PID: 1822
E/AndroidRuntime( 1822): java.lang.NullPointerException: Attempt to invoke virtual method 'void io.flutter.plugin.common.MethodChannel.invokeMethod(java.lang.String, java.lang.Object)' on a null object reference
E/AndroidRuntime( 1822):        at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$EventObserver.onEvent(CloudFirestorePlugin.java:451)
E/AndroidRuntime( 1822):        at io.flutter.plugins.firebase.cloudfirestore.CloudFirestorePlugin$EventObserver.onEvent(CloudFirestorePlugin.java:433)
E/AndroidRuntime( 1822):        at com.google.firebase.firestore.Query.lambda$addSnapshotListenerInternal$2(com.google.firebase:firebase-firestore@@21.3.0:1025)
E/AndroidRuntime( 1822):        at com.google.firebase.firestore.Query$$Lambda$3.onEvent(Unknown Source:6)
E/AndroidRuntime( 1822):        at com.google.firebase.firestore.core.AsyncEventListener.lambda$onEvent$0(com.google.firebase:firebase-firestore@@21.3.0:42)
E/AndroidRuntime( 1822):        at com.google.firebase.firestore.core.AsyncEventListener$$Lambda$1.run(Unknown Source:6)
E/AndroidRuntime( 1822):        at android.os.Handler.handleCallback(Handler.java:873)
E/AndroidRuntime( 1822):        at android.os.Handler.dispatchMessage(Handler.java:99)
E/AndroidRuntime( 1822):        at android.os.Looper.loop(Looper.java:193)
E/AndroidRuntime( 1822):        at android.app.ActivityThread.main(ActivityThread.java:6692)
E/AndroidRuntime( 1822):        at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 1822):        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/AndroidRuntime( 1822):        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
I/Process ( 1822): Sending signal. PID: 1822 SIG: 9
Lost connection to device.
```

**Flutter doctor**
Run `flutter doctor` and paste the output below:

```
[✓] Flutter (Channel stable, v1.12.13+hotfix.9, on Linux, locale en_IN)
    • Flutter version 1.12.13+hotfix.9 at /home/ashutosh/flutter
    • Framework revision f139b11009 (6 weeks ago), 2020-03-30 13:57:30 -0700
    • Engine revision af51afceb8
    • Dart version 2.7.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/ashutosh/Android/Sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /home/ashutosh/Downloads/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Android Studio (version 3.5)
    • Android Studio at /home/ashutosh/Downloads/android-studio
    • Flutter plugin version 41.0.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.44.2)
    • VS Code at /usr/share/code
    • Flutter extension version 3.10.1

[✓] Connected device (1 available)
    • ASUS X01BDA • JBAAGF08C383SR5 • android-arm64 • Android 9 (API 28)

• No issues found!
```

### PR DESCRIPTION
**Pull request by kroikie**
## Description

TL;dr: Remove Android Firestore listeners when the host Activity is destroyed or the engine is detached.

The Cloud Firestore plugins does not currently clean up listeners after the user leaves the app on Android. This results in snapshot listeners sticking around even though associated resources like the channel is no longer available.

When a change occurs to the data, the stranded listeners try to report the updates via the channel that is no longer available. This results in a null pointer exception.

This PR uses lifecycle callbacks of both v1 and v2 embedding to remove listeners before either the app's Activity is destroyed (v1) or the Flutter engine is detached (v2).

## Related Issues

fixes #2517 
fixes #2390 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by ened** [comment]
@kroikie what if developers don’t use cloud firestore in a activity? 

Our app, for example, starts a Android background Service with a built in FlutterEngine for most of the time.

Another example is flutter_workmanager. You could let that one run at any time in the background, update a few fields in firestore, and shutdown.

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java [hunk 1]
```diff
@@ -805,7 +805,7 @@ public void run() {
                   ? MetadataChanges.INCLUDE
                   : MetadataChanges.EXCLUDE;
           listenerRegistrations.put(
-              handle, getQuery(arguments).addSnapshotListener(metadataChanges, observer));
+              handle, getQuery(arguments).addSnapshotListener(activity, metadataChanges, observer));
```
Comment (ened) [inline packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:808]: `activity` will be `null` when the App is run in background only.

It really should be annotated with `@Nullable`.

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java [hunk 2]
```diff
@@ -821,7 +821,8 @@ public void run() {
                   : MetadataChanges.EXCLUDE;
           listenerRegistrations.put(
               handle,
-              getDocumentReference(arguments).addSnapshotListener(metadataChanges, observer));
+              getDocumentReference(arguments)
+                  .addSnapshotListener(activity, metadataChanges, observer));
```
Comment (ened) [inline packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:825]: Same as above `activity` will be `null` when the App is run in background only.
It really should be annotated with `@Nullable`.

**Comment by Salakar** [comment]
The approach we took on the rework we're doing is to store the listener registration instance that is returned from each `addSnapshotListener` call.

Then, when the plugins `onDetachedFromEngine` lifecycle method is called we loop over any stored registrations and call the `remove` method on each registration, thereby cancelling the subscription.

We do this on iOS also.

**Comment by ened** [comment]
@Salakar sounds similar to the approach taken in https://github.com/FirebaseExtended/flutterfire/pull/2717

Looking forward to use the new firestore plugin soon! :-)

**Comment by kroikie** [comment]
@ened Thanks for catching that. I was hoping to avoid the book keeping but the non activity case would indeed not be resolved by this.

@Salakar should I update this PR to use the approach in #2717 or should we wait for the coming fix?

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java [hunk 3]
```diff
@@ -985,6 +996,12 @@ public void onFailure(@NonNull Exception e) {
       this.exception = null;
     }
   }
+
+  private void removeSnapshotListeners() {
+    for (int i = 0; i < listenerRegistrations.size(); i++) {
+      listenerRegistrations.get(i).remove();
+    }
+  }
```
Comment (ened) [inline packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:1004]: Add `listenerRegistrations.clear()`?

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java [hunk 4]
```diff
@@ -76,9 +78,16 @@
...
 
   public static void registerWith(PluginRegistry.Registrar registrar) {
-    CloudFirestorePlugin instance = new CloudFirestorePlugin();
+    final CloudFirestorePlugin instance = new CloudFirestorePlugin();
     instance.activity = registrar.activity();
     instance.initInstance(registrar.messenger());
+    registrar.addViewDestroyListener(new PluginRegistry.ViewDestroyListener() {
+      @Override
+      public boolean onViewDestroy(FlutterNativeView view) {
+        instance.removeSnapshotListeners();
```
Comment (ened) [inline packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:87]: Maybe treat this one as `onDetachedFromEngine` and do all cleanup tasks? @matthew-carroll is that a good assumption?

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java (hunk 3 above)
Comment (kroikie) [reply packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:1004]: Done!

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java (hunk 4 above)
Comment (kroikie) [reply packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java]: @ened are you thinking about cleaning up the channel?

Comment (ened) [reply packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java]: @kroikie it seems good practice (& is part of the V2 generated code), so yes. 

IMHO, this shouldn't be needed in the long run, as a closing engine should clean everything and take everything with it.

Comment (kroikie) [reply packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java]: Agreed

File: packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java [hunk 5]
```diff
@@ -985,6 +999,13 @@ public void onFailure(@NonNull Exception e) {
       this.exception = null;
     }
   }
+
+  private void removeSnapshotListeners() {
+    for (int i = 0; i < listenerRegistrations.size(); i++) {
+      listenerRegistrations.get(i).remove();
```
Comment (ened) [inline packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:1005]: This must use `valueAt(i)`, as listenerRegistrations in a SparseArray (a map-like type) and the key is the handle ID.

Have tested it locally just now and ran into NPE using `.get(i)`.

Comment (kroikie) [reply packages/cloud_firestore/cloud_firestore/android/src/main/java/io/flutter/plugins/firebase/cloudfirestore/CloudFirestorePlugin.java:1005]: Thanks for flagging that!

File: packages/cloud_firestore/cloud_firestore/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,7 @@
+## 0.13.7
+
+* Scope Android listener's to app's Activity.
```
Comment (ditman) [inline packages/cloud_firestore/cloud_firestore/CHANGELOG.md:3]: Is this changelog entry still accurate?

Comment (kroikie) [reply packages/cloud_firestore/cloud_firestore/CHANGELOG.md]: Updated the log!

**Comment by johnvinosh** [comment]
When can I see the update 0.13.7 on pub.dev ?
